code/utils/RandomRange.h

```diff
@@ -261,4 +346,185 @@ UniformRange<Value> parseUniformRange(Value min = std::numeric_limits<Value>::mi
+	inline void reset() {}
+	inline param_type param()
+	{
+		return m_param;
+	}
+	inline void param(param_type p)
+	{
+		m_param = p;
+	}
+	template <typename Generator>
+	inline result_type operator()(Generator& generator, const param_type& param)
+	{
+		if (param.curve < 0) {
+			return 0.f;
+		}
+		float max_integral = Curves[param.curve].GetValueIntegrated(Curves[param.curve].keyframes.back().pos.x);
+		float rand =
+			std::generate_canonical<float, std::numeric_limits<float>::digits, Generator>(generator) * max_integral;
+		float lower_bound = Curves[param.curve].keyframes.front().pos.x;
+		float upper_bound = Curves[param.curve].keyframes.back().pos.x;
+		float curve_min = lower_bound;
+		float curve_max = upper_bound;
+		for (size_t count = 0; count < 16; count++) {
+			float current_pos = (lower_bound + upper_bound) / 2.f;
+			float current_value = Curves[param.curve].GetValueIntegrated(current_pos);
+			if (fl_equal(current_value, rand, max_integral * 0.01f)) {
+				// remap the values to the distribution's range
+				return (current_pos - curve_min) / (curve_max - curve_min) * (param.max - param.min) + param.min;
+			}
+			if (current_value > rand) {
+				upper_bound = current_pos;
+			} else {
+				lower_bound = current_pos;
+			}
+		}
+		// remap the values to the distribution's range
+		return (((lower_bound + upper_bound) / 2.f) - curve_min) / (curve_max - curve_min) * (param.max - param.min) + param.min;
+	}
+	template <typename Generator>
+	inline result_type operator()(Generator& generator)
+	{
+		return this->operator()(generator, m_param);
+	}
+	inline result_type min()
+	{
+		return Curves[m_param.curve].keyframes.front().pos.x;
+	}
+	inline result_type max()
+	{
+		return Curves[m_param.curve].keyframes.back().pos.x;
+	}
+	inline bool operator==(const CurveNumberDistribution& other)
+	{
+		return (m_param.curve == other.m_param.curve && fl_equal(m_param.min, other.m_param.min) && fl_equal(m_param.max, other.m_param.max));
+	}
+	inline bool operator!=(const CurveNumberDistribution& other)
+	{
+		return !(*this == other);
+	}
+};
+
+template <typename _RealType, typename _CharT, typename _Traits>
+std::basic_ostream<_CharT, _Traits>& operator<<(std::basic_ostream<_CharT, _Traits>& __os,
+	const CurveNumberDistribution& __x)
+{
+	__os << __x.param();
+	return __os;
+}
+
+template <typename _RealType, typename _CharT, typename _Traits>
+std::basic_istream<_CharT, _Traits>& operator>>(std::basic_istream<_CharT, _Traits>& __is,
+	const CurveNumberDistribution& __x)
+{
+	int curve;
+	__is >> curve;
+	__x.param(curve);
+	return __is;
 }
+
+using CurveFloatRange = RandomRange<float, CurveNumberDistribution, std::minstd_rand>;
+
+inline CurveFloatRange parseCurveFloatRange(float min = std::numeric_limits<float>::lowest()/2.1f, float max = std::numeric_limits<float>::max()/2.1f) {
+	CurveNumberDistribution::param_type curve_params;
+
+	optional_string("(");
+	stuff_float(&curve_params.min);
+	stuff_float(&curve_params.max);
```

Comment:
The user is just... typing in the curve's min and max? Those should be able to be calculated pretty easily from the curve itself (probably ahead of time and stored on the curve object) rather than allowing for possible incongruity between the actual curve and what the user types in here.

Reply:
These are for remapping the range, after the curve provides its value. (And I'm pretty sure it works correctly, because I tested it and it behaved like I expected.)

I'm absolutely not budging on this one. It's _definitely_ necessary. The median use case here is that you have a handful of common distribution curves, each of which is used several times. If there wasn't a remapping where you're specifying the range, you'd have to make separate curves for all those dozens of cases, go and edit a separate file every time you wanted to tweak your particle velocity or whatever, _come up with names_ for all of them, try to keep track of which ones are for what... It's _so_ much more usable this way.

Reply:
Could they perhaps be optional? I've mentioned before, it's not great to have the same thing defined twice in two different places, there's definite utility here as you've pointed out, but it's easy to imagine a modder who doesn't need to use it and isn't thinking about it being confused by being asked to provide these numbers which may well be the exact same as they already defined in their curve.

Using the curve as is would be the simple, default approach, remapping an existing curve would be the 'advanced' slightly more input approach.

Reply:
Hmm. Okay, yeah, not a bad idea. I can look into that.